File: hw1.py
class Crypto:
    def encrypt(self, text, key):
        raise NotImplementedError("Encrypt method must be implemented")
    
    def decrypt(self, text, key):
        raise NotImplementedError("Decrypt method must be implemented")

class ReverseCipher(Crypto):
    def encrypt(self, text, key=None):
        return text[::-1]
    
    def decrypt(self, text, key=None):
        return text[::-1]

class CaesarCipher(Crypto):
    def encrypt(self, text, key):
        return ''.join(chr((ord(char) - 97 + key) % 26 + 97) if char.islower() else chr((ord(char) - 65 + key) % 26 + 65) if char.isupper() else char for char in text)
    
    def decrypt(self, text, key):
        return ''.join(chr((ord(char) - 97 - key) % 26 + 97) if char.islower() else chr((ord(char) - 65 - key) % 26 + 65) if char.isupper() else char for char in text)

class VigenereCipher(Crypto):
    def encrypt(self, text, key):
        key = key.lower()
        encrypted_text = ""
        key_index = 0
        for char in text:
            if char.isalpha():
                shift = ord(key[key_index % len(key)]) - 97
                encrypted_text += chr((ord(char) - 97 + shift) % 26 + 97) if char.islower() else chr((ord(char) - 65 + shift) % 26 + 65)
                key_index += 1
            else:
                encrypted_text += char
        return encrypted_text
    
    def decrypt(self, text, key):
        key = key.lower()
        decrypted_text = ""
        key_index = 0
        for char in text:
            if char.isalpha():
                shift = ord(key[key_index % len(key)]) - 97
                decrypted_text += chr((ord(char) - 97 - shift) % 26 + 97) if char.islower() else chr((ord(char) - 65 - shift) % 26 + 65)
                key_index += 1
            else:
                decrypted_text += char
        return decrypted_text

def main():
    print("Choose a cipher algorithm:")
    print("1 - Reverse Cipher")
    print("2 - Caesar Cipher")
    print("3 - Vigenere Cipher")
    
    choice = input("Enter choice: ")
    text = input("Enter text: ")
    action = input("Encrypt or Decrypt (e/d): ")
    
    if choice == "1":
        cipher = ReverseCipher()
        key = None
    elif choice == "2":
        key = int(input("Enter key (number): "))
        cipher = CaesarCipher()
    elif choice == "3":
        key = input("Enter key (word): ")
        cipher = VigenereCipher()
    else:
        print("Invalid choice")
        return
    
    if action == 'e':
        print("Encrypted text:", cipher.encrypt(text, key))
    elif action == 'd':
        print("Decrypted text:", cipher.decrypt(text, key))
    else:
        print("Invalid action")

if __name__ == "__main__":
    main()

File: hw1.cpp
#include <iostream>
#include <string>
#include <algorithm>
using namespace std;

class Crypto
{
public:
    virtual string encrypt(const string &text, const string &key) = 0;
    virtual string decrypt(const string &text, const string &key) = 0;
};

class ReverseCipher : public Crypto
{
public:
    string encrypt(const string &text, const string &key) override
    {
        string reversed = text;
        reverse(reversed.begin(), reversed.end());
        return reversed;
    }

    string decrypt(const string &text, const string &key) override
    {
        return encrypt(text, key);
    }
};

class CaesarCipher : public Crypto
{
public:
    string encrypt(const string &text, const string &key) override
    {
        int shift = stoi(key);
        string result;
        for (char ch : text)
        {
            if (isalpha(ch))
            {
                char base = islower(ch) ? 'a' : 'A';
                result += (char)(((ch - base + shift) % 26) + base);
            }
            else
            {
                result += ch;
            }
        }
        return result;
    }

    string decrypt(const string &text, const string &key) override
    {
        int shift = stoi(key);
        return encrypt(text, to_string(26 - shift));
    }
};

class VigenereCipher : public Crypto
{
public:
    string encrypt(const string &text, const string &key) override
    {
        string result;
        int keyIndex = 0;
        for (char ch : text)
        {
            if (isalpha(ch))
            {
                char base = islower(ch) ? 'a' : 'A';
                int shift = tolower(key[keyIndex % key.length()]) - 'a';
                result += (char)(((ch - base + shift) % 26) + base);
                keyIndex++;
            }
            else
            {
                result += ch;
            }
        }
        return result;
    }

    string decrypt(const string &text, const string &key) override
    {
        string result;
        int keyIndex = 0;
        for (char ch : text)
        {
            if (isalpha(ch))
            {
                char base = islower(ch) ? 'a' : 'A';
                int shift = tolower(key[keyIndex % key.length()]) - 'a';
                result += (char)(((ch - base - shift + 26) % 26) + base);
                keyIndex++;
            }
            else
            {
                result += ch;
            }
        }
        return result;
    }
};

int main()
{
    int choice;
    string text, key, action;
    Crypto *cipher = nullptr;

    cout << "Choose a cipher algorithm:" << endl;
    cout << "1 - Reverse Cipher" << endl;
    cout << "2 - Caesar Cipher" << endl;
    cout << "3 - Vigenere Cipher" << endl;
    cin >> choice;
    cin.ignore();

    cout << "Enter text: ";
    getline(cin, text);

    cout << "Encrypt or Decrypt (e/d): ";
    cin >> action;
    cin.ignore();

    switch (choice)
    {
    case 1:
        cipher = new ReverseCipher();
        break;
    case 2:
        cout << "Enter key (number): ";
        cin >> key;
        cin.ignore();
        cipher = new CaesarCipher();
        break;
    case 3:
        cout << "Enter key (word): ";
        cin >> key;
        cin.ignore();
        cipher = new VigenereCipher();
        break;
    default:
        cout << "Invalid choice" << endl;
        return 1;
    }

    if (action == "e")
    {
        cout << "Encrypted text: " << cipher->encrypt(text, key) << endl;
    }
    else if (action == "d")
    {
        cout << "Decrypted text: " << cipher->decrypt(text, key) << endl;
    }
    else
    {
        cout << "Invalid action" << endl;
    }

    delete cipher;
    return 0;
}
